packages/shared/src/server/automations/webhooks.ts

```diff
@@ -0,0 +1,265 @@
+    try {
+      const webhookUrl = new URL(webhookConfig.url);
+      if (webhookUrl.protocol !== "https:") {
+        throw new Error(
+          `Webhook URL must use HTTPS protocol for security. Received: ${webhookUrl.protocol}`,
+        );
+      }
+    } catch (error) {
+      if (error instanceof TypeError) {
+        throw new Error(`Invalid webhook URL: ${webhookConfig.url}`);
+      }
+      throw error;
+    }
+
+    const validatedPayload = PromptWebhookOutboundSchema.safeParse({
+      id: input.executionId,
+      timestamp: new Date(),
+      type: input.payload.type,
+      action: input.payload.action,
+      prompt: input.payload.prompt,
+    });
+
+    if (!validatedPayload.success) {
+      throw new InternalServerError(
+        `Invalid webhook payload: ${validatedPayload.error.message}`,
+      );
+    }
+
+    // Prepare webhook payload with prompt always last
+    const { prompt, ...otherFields } = validatedPayload.data;
+    const webhookPayload = JSON.stringify({
+      ...otherFields,
+      prompt,
+    });
+
+    // Prepare headers with signature if secret exists
+    const requestHeaders: Record<string, string> = {
+      ...WebhookDefaultHeaders,
+      ...webhookConfig.headers,
+    };
+
+    if (!webhookConfig.secretKey) {
+      logger.warn(
+        `Webhook config for action ${automation.action.id} has no secret key, failing webhook execution`,
+      );
+      throw new InternalServerError(
+        "Webhook config has no secret key, failing webhook execution",
+      );
+    }
+
+    if (webhookConfig.secretKey) {
+      try {
+        const decryptedSecret = decrypt(webhookConfig.secretKey);
+
+        const signature = createSignatureHeader(
+          webhookPayload,
+          decryptedSecret,
+        );
+        requestHeaders["Langfuse-Signature"] = signature;
+      } catch (error) {
+        logger.error(
+          "Failed to decrypt webhook secret or generate signature",
+          error,
+        );
+        throw new InternalServerError("Failed to generate webhook signature");
+      }
+    }
+
+    await backOff(
+      async () => {
+        logger.debug(
+          `Sending webhook to ${webhookConfig.url} with payload ${JSON.stringify(
+            webhookPayload,
+          )} and headers ${JSON.stringify(requestHeaders)}`,
+        );
+        const res = await fetch(webhookConfig.url, {
+          method: "POST",
+          body: webhookPayload,
+          headers: requestHeaders,
+        });
+
+        httpStatus = res.status;
+        responseBody = await res.text();
+
+        if (res.status !== 200) {
+          logger.error(
+            `Webhook for project ${projectId} failed with status ${res.status} and response ${responseBody}`,
+          );
+          throw new Error(
+            `Webhook for project ${projectId} failed with status ${res.status}`,
+          );
+        }
+      },
+      {
+        numOfAttempts: 4,
+      },
```

Comment:
This means we have a backoff via the queue _and_ here? Or do we fail the queue on the first attempt? I would consider only relying on the queue here.

Reply:
Ah I see, this is not marked as a recoverable error in the catch block. Might make sense to highlight this in a comment.